### ISSUE TITLE
Optional merging of ctx.state & passed context

### ISSUE THREAD

**Issue by nervgh** [strawbrary/koa-nunjucks-2#18]
Thank you for your work.
## Problems

I use your package with caminte and I have two related issues:
1. in [that](https://github.com/strawbrary/koa-nunjucks-2/blob/v2.1.6/index.js#L64) line the `lodash.merge` can not merge my model because my model uses the prototype' properties https://github.com/biggora/caminte/issues/102
2. Is it really necessary to do deep clone of `this.state`?

``` js
this.state.records = [.../*some data here*/];
yield this.render('path'); // your code will make deep clone in this case
```

I mean that It will be redundant work in this case.
## Possible solutions

I think what [that](https://github.com/strawbrary/koa-nunjucks-2/blob/v2.1.6/index.js#L64) line could look like

``` js
var context = opt_context ? merge({}, this.state, opt_context) : this.state;
```

or

``` js
var context = Object.assign(this.state, opt_context);
```

What do you think about it?

**Comment by nervgh** [comment]
Furthermore, in this line

> ``` js
> ctx.state.records = [.../*some data here*/];
> ```

I have instances of ORM model. These instances have getters which may be long computable (or async). 

### Conclusion

I think copy of `ctx.state` is expensive operation.

**Comment by strawbrary** [comment]
Hello. So I took a look at some other view libraries for Koa and found that the template variable merging strategies are varied - some do deep clones and some do shallow. I think there is a use case for the deep clone + recursive merge so I'm going to keep that in, but what if I added a config option to allow toggling the behavior?

Something like

``` javascript
koaNunjucks({
  recursiveMergeVariables: false,
  path: path.join(__dirname, 'views')
});
```

By default it will use `lodash.merge` but setting `recursiveMergeVariables: false` will cause it to use `Object.assign` instead.

**Comment by nervgh** [comment]
Thanks for your answer.

I think the problem could be split to the next checklist:
- [x] No merge! if `opt_context` is `undefined`. It have to returns `ctx.state` in this case
- [ ] A destination of merging does matter (merging into new object vs merging into current ctx)
- [ ] A strategy of merging does matter (deep vs shallow)

A sample of code:

``` js
function render(view, opt_context, opt_callback) {
  var context = mergeCtx(this.state, opt_context);
  // ...
}

function mergeCtx(ctxState, optContext) {
  if (!optContext) {
    return ctxState;
  }
  // I'm not sure about of code bellow
  return config.useShallowMerging ?
    Object.assign(ctxState, optContext) :
    merge({}, ctxState, optContext);
}
```

**Comment by strawbrary** [comment]
My preference is to create a new object to pass into the Nunjucks renderer to minimize side effects since that can make debugging more difficult. Other than that I think your suggestions are good and I'll add them in soon.

**Comment by strawbrary** [comment]
Just published v2.2.0 with the change. Give it a try.
